### ISSUE TITLE
Closure fires keypress instead of keydown for Firefox

### ISSUE THREAD

**Issue by miketaylr** [google/closure-library#883]
Gecko tried to not fire keypress for non-printable chars in https://bugzilla.mozilla.org/show_bug.cgi?id=968056. But unfortunately that broke Google Docs and had to be backed out (sort of, a pref was landed and it was set to false in Nightly):

https://bugzilla.mozilla.org/show_bug.cgi?id=1442528

The cause appears to be: 

https://github.com/google/closure-library/blob/ccdb3bd5c80094bc65a56053aa3ea782e2711b62/closure/goog/events/keyhandler.js#L282

I'm not sure what the best path forward is right now, especially since the current solution is to check UA strings / versions to know how to behave and we don't have a version with said behavior (because it's unclear what other sites or libraries will break).

**Comment by foolip** [comment]
I'm a little bit confused about what this snippet does. I *think* it will end up matching Chrome, Edge and Safari, which is now the per-spec behavior. If that's right, when this is fixed, then maybe it should be inverted to instead detect Firefox older than a certain version as *having* the bug?

But... what version could be checked, before we know that this change will successfully be made in Firefox?

**Comment by bzbarsky** [comment]
I'm pretty confused by the Closure code.  Firefox fires keydown for non-printable chars.  That's not changing.  The only change being made is to stop firing keypress for some of them.

Put another way, what would break if this code were changed to:

    goog.events.KeyHandler.USES_KEYDOWN_ = true;

?

**Comment by birtles** [comment]
cc @masayuki-nakano

**Comment by miketaylr** [comment]
There's a similar condition in actionhandler.js as well:

https://github.com/google/closure-library/blob/be0326ba32f47ea74af4fbe8df038e579ec35215/closure/goog/events/actionhandler.js#L84-L89

**Comment by masayuki-nakano** [comment]
Although I've not read the code around key event handling, I have no idea why the original developer needed to listen to `keypress` events only for Firefox because `keypress` events for non-printable keys and key combinations which are fired only on Firefox  are just redundant. So, I mean, any web applications must not need to *wait* keypress event only on Firefox. Instead, web apps should/can listen to `keydown` event to handle non-printable key actions.

But if there were other compatibility issue which made the original developer listen to `keypress` for Firefox, we (Mozilla) need to fix the incompatible points first.

So, I think that if there were no reasons to listen to `keypress` for Firefox, it just does not make sense to set `USES_KEYDOWN_` to false.

**Comment by dbaron** [comment]
The way the code is structured feels like using keydown was a workaround for the "bug" of not firing keypress in the non-Gecko browsers.

**Comment by masayuki-nakano** [comment]
Hmm, if so, there should be KeyboardEvent.isPrintable or something which actually indicates if followed by keypress event... (I proposed that but rejected.)

**Comment by masayuki-nakano** [comment]
Reading around `USES_KEYDOWN_` briefly. Looks like that `firesKeyPressEvent` should be fixed too. It checks UA string to *guess* if the `keydown` event is followed by `keypress` event. Additionally,  looks like that there should be a check of ignoring unexpected `keypress` events (i.e., if `firesKeyPressEvent` returns false for a `keypress` event, it should've already dispatched closure's event, so, not necessary to dispatch same event again.).

**Comment by dtapuska** [comment]
@EiraGe @NavidZ do either of you want to take a stab at updating closure and testing it against FireFox?

**Comment by ojanvafai** [comment]
This code was written in 2008. At the time, keydown events didn't repeat in Firefox when you held down a key. Has that changed? If so, USES_KEYDOWN_ should change to whatever Firefox version changed this behavior or (even better) be removed entirely.

**Comment by dtapuska** [comment]
Appears this bug is still open.. https://bugzilla.mozilla.org/show_bug.cgi?id=91592
I do get multiple keydown's for FireFox on Windows 10, and Debian. @masayuki-nakano please advise.

**Comment by bzbarsky** [comment]
> I do get multiple keydown's for FireFox on Windows 10, and Debian for held keys correctly.

I ge multiple keydowns on Mac as well, for Firefox versions starting with Firefox 4 (Firefox 3.6 and earlier do not fire multiple keydowns for a held key).

Looking a bit, I think the Mac behavior was changed to the current one in https://bugzilla.mozilla.org/show_bug.cgi?id=599887 and the Linux behavior in https://bugzilla.mozilla.org/show_bug.cgi?id=597981 (though that was just dealing with a change in how GTK sends key events; before GTK made that change Firefox on Linux would send repeated keydown events).

Windows repeated keydown all along' see https://bugzilla.mozilla.org/show_bug.cgi?id=91592#c8 from back in 2004.

So I think bug 91592 was fixed by the fixes in bug 597981 and bug 599887, and those fixes shipped in Firefox 4.

**Comment by masayuki-nakano** [comment]
Yeah, I fixed a lot around the key repeat behavior. Here is the document:
https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent#Auto-repeat_handling

When I implement KeyboardEvent.repeat on Firefox, I checked all platforms which currently Firefox supports. In any those platforms, I see same behavior except some special cases like OS doesn't use auto-repeat, etc.
